README.md
# nixieclock

A nixietube desktop clock. Runs on Mac, Linux, Windows.

![Image of nixieclock](https://raw.githubusercontent.com/erco77/nixieclock/main/docs/images/nixiedate-anim.gif)

The nixieclock program is a simple little desktop clock that shows the
current time using a 'nixie tube' style display, and can display the
time in various time formats as set by the `-s` flag.

It can be run from the command line with `./nixieclock` (or on windows, `nixieclock.exe`), and moved around on the screen by click and dragging it to a new position. When closed and re-opened, it remembers the last position and time format you configured last.

You can choose different time formats by right clicking, and choose from different presets:
![Popup Menu](https://raw.githubusercontent.com/erco77/nixieclock/main/docs/images/nixieclock-popup-menu.png)

..or use the command line `-s` flag to create custom time formats using strftime(3) format strings, e.g.

Command | Result
------- | -------
`nixieclock -s "%H:%M %m/%d/%y"` | ![Time format example](https://raw.githubusercontent.com/erco77/nixieclock/main/docs/images/nixiedate-digitformat.jpg)
`nixieclock -s "%H:%M %a %m/%d/%y"` | ![Time format example](https://raw.githubusercontent.com/erco77/nixieclock/main/docs/images/nixiedate-hm_dow_mm_dd_yy.jpg)
`nixieclock -s "%a %b %m %H:%M:%S %Y"` | ![Time format example](https://raw.githubusercontent.com/erco77/nixieclock/main/docs/images/nixiedate-unix-format.jpg)

For more info, see the `strftime(3)` and `nixieclock(1)` man pages, or the `docs/nixieclock.html` html documentation, or invoke nixieclock with the `-help` flag.

Also, there is also a text file viewer using nixie digits; try running:
`./nixie-text-viewer` The Up/Down arrow keys move through the source code
a line at a time, PgUp/PgDn to move a page at a time, Home/End to move
quickly to top/bottom of file.

The '`NixieDisplay`' class that can be used in your own applications to
do general display of multiline ASCII text.

For instance, the '`nixie-text-viewer`' demo shows the program's own source
code in a text viewer using the same nixie font used by the clock..
![Nixie text viewer screenshot](https://raw.githubusercontent.com/erco77/nixieclock/main/docs/images/nixie-text-viewer.jpg)

## LICENSE

This code is licensed as GPL3; see `LICENSE` file for details.

## BUGS

    To report bugs, create an 'issue' on the project's github page:
    https://github.com/erco77/nixieclock

## DOCUMENTATION

See `./docs/nixieclock.html`

## WEBSITE

    github:  https://github.com/erco77/nixieclock
    webpage: http://seriss.com/people/erco/unixtools/nixieclock

## CONTRIBUTORS

* Greg Ercolano - nixieclock application (first written for the SGI + Linux, later OSX and Windows)
* Vance Kochenderfer - man page
* Lemon - freebsd and countdown patch
* Terry mrtembryyahoocom - mods for mingw linux cross compile, fixes


CHANGES.txt
1.66 -- Convert to git/github, GPL2 -> GPL3

1.65 -- ??.??.??

    o Applied Alexander Clouter's #include and Makefile mods

1.64 -- 04/10/12

    o Restored lemon's '-c <epoch>' flag to enable 'beer o'clock' countdown
      (See -help and man page for examples of how to use)

    o Added the -N flag to help the -c <epoch> flag from affecting
      the ~/.nixieclock file.

    o tar file now has version number in the extraction
      (eg. extracts as "./nixieclock-<VERSION#>/*")

1.63 -- 03/15/12 [INTERNAL RELEASE]
    o Changed 3dsite.com -> seriss.com

    o Added 'ENVIRONMENT' section to docs re: TZ variable (for Terry).
      Example shows how to use TZ to show the time in GMT.

    o Added small fixes to support mingw from Terry

    o FLTK added to help screen (LGPL/FLTK license compliance)

1.62 -- 11/28/11 [INTERNAL RELEASE]
    o Fix for Daniel Ott regarding Fl::args() parsing issue;
      parse_args() wasn't returning word count.

    o Added docs for -m (flag was recognized but not documented)

    o Added -D flag: same as -d, but with 4 digit year (13:59 12/31/1999)

    o Removed some extraneous code

1.61 -- 06/08/11

    o Fixed a silly bug that made clock redraw itself waaay too often
      and used too much cpu

1.60 -- 05/05/11

    o Added right click popup menu to set various time formats.
      For macs with single-button mice, Ctrl-Click will popup the menu.

    o Clock settings now saved in a ~/.nixieclock file
      so that e.g. clock will come up in same position 
      it was last moved to, and same time format it was last
      set to.

    o Fixed command line help menu

    o When time formats are adjusted, clock maintains
      distance to closest edge of screen.

    o Added '%', '&' and '|' characters to NixieDisplay.

    o nixie-text-viewer: added Home and End keys.
      Added tooltip to show supported navigation keys.

    o (INTERNAL) Reorg'ed clock code into a 'NixieClock' class
      in order to implement popup menu features.

    o (INTERNAL) Reformatted code for 'fltk style' braces
      which makes code a bit shorter for easy reading.

    o (INTERNAL) Had to rename nixieclock.C -> nixieclock-app.C
      to avoid conflict with NixieClock vs. nixieclock on platforms
      that don't care about case >:/

1.51 -- 02/10/10

    o Added #include <stdio.h> to top of nixie-text-display.C 
      to fix compile problem reported by Jozef Riha.

1.50 -- 11/15/09

    o Applied Erkki Myllynen's 11/15/09 fixes/suggs
         o fl_draw_image() -1 problem
	 o Digit.C y>=0 problem
	 o "6" digit repainted/shifted right 1 bit
	 o -Wno-write-strings

    o Fixed '9'

    o Added nixie-text-display to default build

1.49 -- 10/27/09

    o For OSX, made 'nofork' default to on, to prevent these errors
      when running on OSX 10.5 and up:

	The process has forked and you cannot use this CoreFoundation functionality safely. You MUST exec().
	Break on __THE_PROCESS_HAS_FORKED_AND_YOU_CANNOT_USE_THIS_COREFOUNDATION_FUNCTIONALITY___YOU_MUST_EXEC__() to debug.
	The process has forked and you cannot use this CoreFoundation functionality safely. You MUST exec().
	Break on __THE_PROCESS_HAS_FORKED_AND_YOU_CANNOT_USE_THIS_COREFOUNDATION_FUNCTIONALITY___YOU_MUST_EXEC__() to debug.

1.48 -- 06/26/09

    o Fixed problem causing unnecessary cpu use;
      update rate was too fast under normal circumstances

1.47 -- ??/??/06

    o Makefile converted to flkt-config

    o Moved docs into docs

    o Moved icons into icons

    o Created nixie-test-viewer

    o Readme modified

1.46 -- 04/01/06

    o Made height smaller to match older version of nixieclock

    o NixieDisplay standalone now supports multiline text.
      (Try 'make NixieDisplay ; ./NixieDisplay')

1.45 -- 01/01/06

    o Makefile fixes: nixieclock.app, cleanup bin.XXX files,
      VERSION number now comes from first entry in CHANGES file

    o Darwin: strip(1) now run before Rez

    o Darwin: Detects FLTK 1.1.7, enables no border 
      (noborder problem was fixed in fltk 1.1.7 for OSX)

    o Darwin: now starts in lower right by default
      (like other platforms)

    o Improved help

1.44 -- 12/15/05

    o Windows version now opens in lower right by default

    o Fixed behavior of -d option under Windows

    o Rewrite of code to be more modular, fade code in progress

    o Redo of Makefiles (sorry, I backed off from fltk-config)

1.43 -- 11/12/03

    o Added an icon for windows; changes made to Makefile for resource compiler

    o Command line help now comes up in a 'dialog window' under Windows

    o Now using fltk-config to compile programs

    o Makefile now automatically chooses Makefile.<PLATFORM>

1.42 -- 08/08/03

    o Added -c from lemon at aldigital dot co dot uk

1.41 -- 02/05/02

    o Efficient redraws/holdtime now .5 secs

1.40 -- 02/02/02

    o Rewrote using fltk image drawing, cleanup

1.30 -- 07/28/01

    o Added entire nixie alphabet

1.20 -- 07/19/01

    o OPENGL version, renamed nixiedate

1.10 -- 11/25/00

    o added composite digits


nixieclock-app.C
/************************************************************************

    nixieclock - a nixie desktop timepiece
    Copyright (C) 2000,2020 Greg Ercolano, erco@seriss.com

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

    To report bugs, create an 'issue' on the project's github page:
    https://github.com/erco77/nixieclock

*************************************************************************/

#include <stdio.h>	// stderr
#include <unistd.h>	// fork()
#include <FL/Fl.H>	// Fl::error()
#include <FL/fl_ask.H>	// Fl::error()
#include "NixieClock.H"
#include "version.H"

// GLOBALS
NixieClock *G_nixieclock = 0;		// main window
int         G_nofork = 0;		// set if we shouldn't fork

void HelpAndExit() {
    Fl::error(
        "Nixieclock -- Version " VERSION "\n"
	"(C) Copyright Greg Ercolano 2000, 2020. (GPL V3 LICENSE)\n"
	"This program uses FLTK, a GUI library available from fltk.org\n"
	"\n"
	"USAGE\n"
	"    nixieclock [options]\n"
	"\n"
	"OPTIONS\n"
	"    -m              # month oriented date (default, eg. 07:57 Sat Feb 07)\n"
	"    -d              # digit date mode (eg. 07:57 07-30-01)\n"
	"    -D              # digit date mode (eg. 07:57 07-30-2001)\n"
	"    -s <timestr>    # use strftime(3) date format\n"
	"    -t              # test pattern for alphanumeric display\n"
	"    -n              # no background mode (don't fork)\n"
	"    -geometry <pos> # sets geometry (x windows style positions)\n"
	"    -c <epochtime>  # counts down to epochtime\n"
	"    -N              # no save: disables saving the current settings to ~/.nixieclock\n"

// NOT YET
//	"    -r cmd [args]   # runs command, each line of output updates in clock\n"
	"    -a              # above all windows (not supported on all platforms)\n"
	"    -b              # enable window borders\n"
	"    -nb             # disable window borders (default)\n"
	"    -v              # enables verbose mode/debugging\n"
	"    -h              # help\n"
	"\n"
	"FILES\n"
	"    Previous settings are saved in ~/.nixieclock\n"
	"\n"
        "BUGS\n"
	"    To report bugs, create an 'issue' on the project's github page:\n"
        "    https://github.com/erco77/nixieclock\n"
	"\n"
	"EXAMPLES\n"
	"    nixieclock                               -- (uses last set date format. Default: -m)\n"
	"    nixieclock -m                            -- 07:57 SAT FEB 07 (default)\n"
	"    nixieclock -d                            -- 07:57 07-30-01\n"
#ifdef _WIN32
	"    nixieclock -s %%c                         -- Locale specific: 01/01/00 23:59:59\n"
#else
	"    nixieclock -s %%c                         -- Locale specific: MON JAN 01 23:59:59 2000\n"
#endif
	"    nixieclock -s \"%%I:%%M %%p %%m/%%d/%%y\"        -- 12:59 PM 12/31/00\n"
	"    nixieclock -s \"%%H:%%M:%%S\"                 -- 23:59:59 (counts seconds)\n"
	"    nixieclock -s \"%%H:%%M:%%S %%a %%b %%d\"        -- 23:59:59 SAT DEC 31\n"
	"    nixieclock -s \"%%a %%b %%m %%H:%%M:%%S %%Y\"     -- Unix format; MON JAN 01 23:59:59 2000\n"
	"    nixieclock -s \"%%a %%b %%m %%H:%%M:%%S %%Y %%Z\"  -- Above + timezone; MON JAN 01 23:59:59 2000 PDT\n"
	"    nixieclock -geometry -0-0                 -- position lower right\n"
	"    nixieclock -geometry -5-5                 -- position 5 pixels lower right\n"
	"    nixieclock -geometry +0-0                 -- position lower left\n"
	"    nixieclock -N -c `perl -e 'print time()+3600;'` -s \"BEER: %%H:%%M:%%S\"\n"
	"                                              -- 'beer o'clock' countdown timer\n"
	"                                                  Sets 'countdown timer' to 1hr from now\n"
	"                                                  (3600 secs=1hr)\n"
// NOT YET
//	"    nixieclock -r 'vmstat 3' -geom 1280x40   -- vmstat appears in clock\n"
	"\n");

    exit(1);
}

#define ISIT(val)	if ( strcmp(argv[t], val) == 0 )

// Handle parsing our app's own arguments
int parse_args(int argc, char **argv, int &t) {
    int parsed_words = 0;
    ISIT("-nb") {
	++t;
	G_nixieclock->SetBorder(0);
	parsed_words++;
    } else ISIT("-b") {
	++t;
	G_nixieclock->SetBorder(1);
	parsed_words++;
    } else ISIT("-n") {
	++t;
	G_nofork = 1;
	parsed_words++;
    } else ISIT("-N") {
	++t;
	G_nixieclock->SetNoSave(1);
	parsed_words++;
    } else ISIT("-a") {
	++t;
	G_nixieclock->SetAbove(1);
	parsed_words++;
    } else ISIT("-s") {
	if ( ++t >= argc ) {
	    fprintf(stderr, "-s: missing argument\n");
	    exit(1);
	}
	G_nixieclock->SetTimeString(argv[t]);
	G_nixieclock->SetTickSpeed(0.25);	// fast update (assume user specified time that includes seconds)
	++t;
	parsed_words += 2;
    } else ISIT("-c") {
	if ( ++t >= argc ) {
	    fprintf(stderr, "-c: missing argument (expected a time(2) value)\n");
	    exit(1);
	}
	unsigned long ulval = 0;
	if ( sscanf(argv[t], "%lu", &ulval) != 1 ) {
	    fprintf(stderr, "-c: unexpected non-integer '%s'\n", argv[t]);
	    exit(1);
	}
	G_nixieclock->SetCountdown((time_t)ulval);
	++t;
	parsed_words += 2;
    } else ISIT("-h") {
	// 23:59
	++t;
	G_nixieclock->SetTimeString("%H:%M");
	G_nixieclock->SetTickSpeed(29.0);
	parsed_words++;
    } else ISIT("-d") {
	// 23:59 12/31/99
	++t;
	G_nixieclock->SetTimeString("%H:%M %m/%d/%y");
	G_nixieclock->SetTickSpeed(29.0);
	parsed_words++;
    } else ISIT("-D") {
	// 23:59 12/31/1999
	++t;
	G_nixieclock->SetTimeString("%H:%M %m/%d/%Y");
	G_nixieclock->SetTickSpeed(29.0);
	parsed_words++;
    } else ISIT("-m") {
	// 23:59 Sat Dec 31
	++t;
	G_nixieclock->SetTimeString("%H:%M %a %b %d");
	G_nixieclock->SetTickSpeed(29.0);
	parsed_words++;
    } else ISIT("-t") {
	++t;
	G_nixieclock->SetTimeString("ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789~!@#$%%^&*()_+`[]{}\\|;':\",./<>?");
	G_nixieclock->SetTickSpeed(30.0);
	parsed_words++;
    } else ISIT("-v") {
	++t;
	G_nixieclock->SetVerbose(1);
	parsed_words++;
    } else ISIT("-geometry") {
        // Check for, but don't parse this one.. let FLTK handle it
	G_nixieclock->SetGeometryForce(1);
	//parsed_words++;
    }
    return(parsed_words);
}

// MAIN
int main(int argc, char **argv) { 
    // BANNER
    fprintf(stderr, "\nnixieclock -- Version %s\n"
                    "(C) Copyright 2011,2020 Greg Ercolano (erco@seriss.com).\n"
		    "This code is GPL V3.\n\n", VERSION); 
    
    // PARSE HELP
    for ( int t=1; t<argc; t++ ) {
	// HELP
	if ( strcmp(argv[t], "-h") == 0 ||
	     strcmp(argv[t], "-help") == 0 ||
	     strcmp(argv[t], "--help") == 0 ) {
	     HelpAndExit();
	}
    }

    // MAKE MAIN WINDOW
    //    Arbitrary size/position for init;
    //    SetTimeString() will determine actual size/position.
    //
    G_nixieclock = new NixieClock(Fl::w()-30,Fl::h()-30,25,25,"nixieclock");
    G_nixieclock->ForceUpperCase(1);
    G_nixieclock->SetUpdateSpeed(0.50);
    G_nixieclock->SetBorder(0);

    // Load previously saved settings
    //    Then command line arguments can alter from there..
    //
    if ( G_nixieclock->LoadSettings() < 0 ) {
        fprintf(stderr, "%s: couldn't load previous settings: %s\n",
	    argv[0], G_nixieclock->GetErrmsg());
	fl_alert("%s: couldn't load previous settings: %s", argv[0], G_nixieclock->GetErrmsg());
    }

    // PARSE ARGUMENTS
    //    Handles both our own args and FLTK's.
    //    Our arguments are handled by the parse_args() function above.
    //
    int i = 1;
    if ( Fl::args(argc, argv, i, parse_args) < argc ) {
        fprintf(stderr, "%s: unknown option '%s'\n", argv[0], argv[i]);
	exit(1);
    }
#ifdef __APPLE__
    // Do not fork on mac; we loose mach ports bindings (or whatever)
    G_nofork = 1;
#else
    if ( G_nofork == -1 ) G_nofork = 0;
#endif

#ifndef _WIN32
#ifndef DARWIN
    // UNIX: RUN IN THE BACKGROUND
    if ( ! G_nofork ) {
	switch ( fork() ) {
	    case -1: perror("fork"); exit(1); 	// ERROR
	    case  0: break;			// CHILD
	    default: exit(0);			// PARENT DISCONNECTS
	}
    }
#endif /*DARWIN*/
#endif /*_WIN32*/

    // SET VISUAL MODE
    //
    //     FL_RGB prevents color map eating under X windows.
    //
    //     FL_DOUBLE|FL_INDEX enables hardware double buffering,
    //     but looks bad on Indy; digit backgrounds are gray.
    //     Fltk does a good job emulating double buffering
    //     in Fl_Double_Window w/out flickering, so leave off.
    //
    Fl::visual(FL_RGB);

    //// BAD FOR WINDOWS (causes clock to appear in upperleft until "touched")
    //// G_nixieclock->free_position();

    G_nixieclock->show(argc, argv);

    // HACK: Need to call this again /after/ show
    //       for it to work on win32..
    if ( G_nixieclock->GetAbove() ) {
        G_nixieclock->SetAbove(1);
    }

    if ( G_nixieclock->SaveSettings() ) {	// save as new settings
        fprintf(stderr, "%s: couldn't save previous settings: %s\n",
	    argv[0], G_nixieclock->GetErrmsg());
	fl_alert("%s: couldn't save previous settings: %s", 
	    argv[0], G_nixieclock->GetErrmsg());
    }
    return(Fl::run());
}
